### PR TITLE
Escape environment vars in the default rarun profile

### PR DESCRIPTION
**Pull request by abcSup**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**
Escape the environment variable values with `'` in the startup rarun profile to prevent misinterpreting special characters as rarun prefixes.

For example, an env var with special characters like `DISPLAY=:0` should not be parsed as a hexpair string by `run.c:getstr`

**Test plan**

...

**Closing issues**

### CODE REVIEW

**Comment by karliss** [comment]
What if the environment variable contains `'` symbol or any other special symbol interpreted by rarun profile mechanism. From what I understood `r_run_parseline` uses run.c/getstr which deals s with a few special symbols.

**Comment by abcSup** [comment]
> What if the environment variable contains `'` symbol or any other special symbol interpreted by rarun profile mechanism. From what I understood `r_run_parseline` uses run.c/getstr which deals s with a few special symbols.

You are right. I am trying to copy all original env vars from the environment as they are exactly and not interpret them as special characters. For example, `:0` in a line `setenv=DISPLAY=:0` will be parsed by `getstr` as a hexpair string which ends up setting `DISPLAY` to `"\0"`, so I escape all original env vars from the environment with `'` to prevent `getstr` from interpreting them differently.

**Comment by karliss** [comment]
As a general rule of thumb working string escaping system need at least these 3 parts:
* quotes
* mechanism to escape quotes
* mechanism to escape the escape mechanism

If you have only first one or first two it can't be a working string escape system which supports all symbols. In your code I only see the first one. Without testing this makes me think that current code doesn't properly solve the problem.

In some use cases the whole problem can be sidestepped if there are string bounds established using external information or some symbol that can never appear in the text (with limitation that this symbol can't be used in the text).

**Comment by karliss** [comment]
In this case it seems that `'` doesn't cause problem because parser tries to sidestep the problem by using end of line for locating end of string. This means that single quotes don't support newline symbol. Just tested using environment variable that contains newline and it indeed broke the parser. This is reasonable compromise for hand written rarun profiles but not good enough when we create one automatically from existing environment which could contain anything.

Rarun also treats doublequotes `"` differently. It uses `r_str_unescape` looks. Just need to choose correct `r_str_escape_*` method.

**Comment by abcSup** [comment]
Looking at the code, `getstr` returns the exact string if it is wrapped between single quotes `'` and does not try to interpret single quotes or other special symbols inside that string again. I find that wrapping the original values with single quotes `'` is enough and there is no need to further escaping any special symbols inside the env variable value.

I understand that you are trying to teach me the proper way to deal with this. What I mean by "enough" is not that I try to ignore what you just said and insist on my own way, but I tried and cannot think of a better way to handle this because there seems to be no way a malicious string could get away with if I just escape it with `'` from the code in `run.c`.

**Comment by abcSup** [comment]
> This means that single quotes don't support newline symbol. Just tested using environment variable that contains newline and it indeed broke the parser. This is reasonable compromise for hand written rarun profiles but not good enough when we create one automatically from existing environment which could contain anything.

Ok, I see it now.

File: libr/main/radare2.c [hunk 1]
```diff
@@ -391,7 +391,17 @@ R_API int r_main_radare2(int argc, const char **argv) {
 	if (e) {
 		RStrBuf *sb = r_strbuf_new (NULL);
 		while (e && *e) {
-			r_strbuf_appendf (sb, "setenv=%s\n", *e);
+			char *k = strdup (*e);
```
Comment (trufae) [inline libr/main/radare2.c:394]: may be worth moving this logic into a separate function

File: libr/main/radare2.c [hunk 2]
```diff
@@ -391,7 +391,17 @@ R_API int r_main_radare2(int argc, const char **argv) {
 	if (e) {
 		RStrBuf *sb = r_strbuf_new (NULL);
 		while (e && *e) {
```
Comment (trufae) [inline libr/main/radare2.c:393]: e should not be null, this while can be just `while (*e)`

**Review by trufae — DISMISSED** [review]

File: libr/main/radare2.c [hunk 3]
```diff
@@ -391,7 +391,17 @@ R_API int r_main_radare2(int argc, const char **argv) {
...
-			r_strbuf_appendf (sb, "setenv=%s\n", *e);
+			char *k = strdup (*e);
+			char *v = strchr (k, '=');
+			if (v) {
+				*v++ = 0;
+				v = r_str_escape (v);
```
Comment (karliss) [inline libr/main/radare2.c:398]: `r_str_escape` strips anything that look like terminal color codes. 

Weird I thought I made this comment twice already but it wasn't here.

Comment (abcSup) [reply libr/main/radare2.c:398]: Thanks again and sorry for sounding a bit rude previously as I did not think of the scenario of having newline as values in env vars. I will look into this again.

**Comment by trufae** [comment]
ping

File: libr/main/radare2.c (hunk 1 above)
Comment (abcSup) [reply libr/main/radare2.c]: I have moved it to a new function called r_run_get_environ_profile

File: libr/main/radare2.c (hunk 3 above)
Comment (abcSup) [reply libr/main/radare2.c]: I have tested color codes and escape sequence and it should work now.

**Review by radare — APPROVED** [review]
+1!

**Review by karliss — APPROVED** [review]

**Comment by ret2libc** [comment]
@XVilka I did not follow this changes, sorry. If you are ok with them, feel free to merge, there are already 2 approvals.

**Pull request merged**